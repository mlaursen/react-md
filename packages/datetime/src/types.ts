/**
 * @since 6.4.0
 */
export type ReadonlyDate = Readonly<Date>;

/**
 * @since 6.4.0
 */
export interface OptionsWithDate {
  date: ReadonlyDate;
}
