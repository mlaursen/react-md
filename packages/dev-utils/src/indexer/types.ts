import { ReactNode } from "react";

/**
 * An anchor that can be used for a table of contents.
 */
export interface TOCAnchor {
  /**
   * This should be a string that can be used to link to another part of a page.
   * For example: `#this-is-a-title-id`
   */
  anchor: string;

  /**
   * The title for the anchor. Note: the `anchor` is normally generated from the
   * title's value.
   */
  title: string;
}

/**
 * The metadata for a demo on one of the package's demo pages.
 */
export interface DemoMetadata {
  /**
   * The title for the demo.
   */
  title: string;

  /**
   * The description for the demo. This is generated based on the demo's
   * markdown file that is normally provided for each demo as the `description`
   * field..
   */
  summary: string;
}

export type MetadataType =
  | "blog"
  | "guide"
  | "theme"
  | "sassdoc"
  | "demos"
  | "demo"
  | "api"
  | "changelog";

/**
 * A record of all the table of content anchors based on pathname for quick
 * lookups.
 */
export type TOCRecord = Record<string, readonly TOCAnchor[] | undefined>;

/**
 * This is used for searching within the documentation site.
 */
export interface RouteMetadata {
  /**
   * The title of the "thing" that can be directly navigated to.
   */
  title: string;

  /**
   * A short summary describing what this "thing" is. This is normally generated
   * from related markdown files.
   */
  summary: string;

  /**
   * This is the url that should be used by nextjs since some of the pages are
   * dynamically generated with `[id]`s in the url.
   */
  pageUrl: string;

  /**
   * The pathname for this item. This is what will be displayed in the browser
   * url.
   */
  pathname: string;

  /**
   * The type of the metadata item.
   */
  type: MetadataType;
}

export interface IndexedResult {
  readonly tocs: TOCRecord;
  readonly metadata: readonly RouteMetadata[];
}

export interface RouteNavItem {
  href: string;
  children: string;
  leftAddon?: ReactNode;
  routes?: NavItem[];
}

export interface DividerNavItem {
  divider: true;
  parentId?: string;
}

export interface SubheaderNavItem {
  subheader: true;
  parentId?: string;
  children: string;
}

export type NavItem = RouteNavItem | DividerNavItem | SubheaderNavItem;
