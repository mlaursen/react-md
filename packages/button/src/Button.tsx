/* eslint-disable react/button-has-type */
import React, { ButtonHTMLAttributes, forwardRef, ReactNode } from "react";
import {
  InteractionStatesOptions,
  useInteractionStates,
} from "@react-md/states";
import { PropsWithRef } from "@react-md/utils";

import buttonThemeClassNames, {
  ButtonThemeProps,
} from "./buttonThemeClassNames";
import FAB, { FABPosition, FABProps } from "./FAB";

/**
 * This interface includes all the props that the `Button` component accepts so
 * the main use case might be creating a functionality wrapper for the `Button`
 * component, but passes all props down as normal.
 */
export interface ButtonProps
  extends ButtonThemeProps,
    ButtonHTMLAttributes<HTMLButtonElement>,
    Omit<InteractionStatesOptions<HTMLButtonElement>, "disableSpacebarClick"> {
  /**
   * The button's type attribute. This is set to "button" by default so that
   * forms are not accidentally submitted when this prop is omitted since
   * buttons without a type attribute work as submit by default.
   */
  type?: "button" | "reset" | "submit";

  /**
   * Any children to render within the button. This will normally just be text
   * or an icon.
   *
   * Please note that it is considered invalid html to have a `<div>` as a
   * descendant of a `<button>`.
   */
  children?: ReactNode;

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FABPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FABProps, HTMLSpanElement>;
}

/**
 * The `Button` component is a simple wrapper for the `<button>` element that
 * defaults the `type` attribute to `"button"` so that it does not automatically
 * submit forms by default. It also supports multiple themes, rendering as an
 * icon button, or even as a floating action button.
 *
 * The default theme will be a clear text button unless the `floating` prop has
 * been provided where it will render as an icon button by default instead.
 */
const Button = forwardRef<HTMLButtonElement, ButtonProps>(function Button(
  {
    type = "button",
    disabled = false,
    floating = null,
    floatingProps,
    theme = floating ? "secondary" : "clear",
    themeType = floating ? "contained" : "flat",
    buttonType = floating ? "icon" : "text",
    className: propClassName,
    children,
    disableRipple,
    disableProgrammaticRipple,
    rippleTimeout,
    rippleClassNames,
    rippleClassName,
    rippleContainerClassName,
    enablePressedAndRipple: propEnablePressedAndRipple,
    disablePressedFallback,
    onClick,
    onKeyUp,
    onKeyDown,
    onMouseUp,
    onMouseDown,
    onMouseLeave,
    onTouchStart,
    onTouchMove,
    onTouchEnd,
    ...props
  },
  ref
) {
  const enablePressedAndRipple =
    typeof propEnablePressedAndRipple === "boolean"
      ? propEnablePressedAndRipple
      : themeType === "contained";
  const propHandlers = {
    onClick,
    onKeyUp,
    onKeyDown,
    onMouseUp,
    onMouseDown,
    onMouseLeave,
    onTouchStart,
    onTouchMove,
    onTouchEnd,
  };

  const isDisabledTheme = theme === "disabled";
  const { ripples, className, handlers } = useInteractionStates({
    handlers: propHandlers,
    className: buttonThemeClassNames({
      theme,
      themeType,
      buttonType,
      disabled,
      className: propClassName,
    }),
    disabled: disabled || isDisabledTheme,
    disableRipple,
    disableProgrammaticRipple,
    disablePressedFallback,
    rippleTimeout,
    rippleClassNames,
    rippleClassName,
    rippleContainerClassName,
    enablePressedAndRipple,
  });

  return (
    <FAB position={floating} {...floatingProps}>
      <button
        aria-disabled={isDisabledTheme || undefined}
        {...props}
        {...(isDisabledTheme ? undefined : handlers)}
        ref={ref}
        type={type}
        className={className}
        disabled={disabled}
      >
        {children}
        {ripples}
      </button>
    </FAB>
  );
});

if (process.env.NODE_ENV !== "production") {
  try {
    const PropTypes = require("prop-types");

    Button.propTypes = {
      type: PropTypes.oneOf(["button", "reset", "submit"]),
      className: PropTypes.string,
      theme: PropTypes.oneOf([
        "clear",
        "primary",
        "secondary",
        "warning",
        "error",
        "disabled",
      ]),
      themeType: PropTypes.oneOf(["flat", "outline", "contained"]),
      buttonType: PropTypes.oneOf(["text", "icon"]),
      disabled: PropTypes.bool,
      children: PropTypes.node,
      disableRipple: PropTypes.bool,
      disableProgrammaticRipple: PropTypes.bool,
      disablePressedFallback: PropTypes.bool,
      rippleTimeout: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.shape({
          appear: PropTypes.number,
          enter: PropTypes.number,
          exit: PropTypes.number,
        }),
      ]),
      rippleClassNames: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          appear: PropTypes.string,
          appearActive: PropTypes.string,
          enter: PropTypes.string,
          enterActive: PropTypes.string,
          enterDone: PropTypes.string,
          exit: PropTypes.string,
          exitActive: PropTypes.string,
          exitDone: PropTypes.string,
        }),
      ]),
      rippleClassName: PropTypes.string,
      rippleContainerClassName: PropTypes.string,
      enablePressedAndRipple: PropTypes.bool,
      floating: PropTypes.oneOf([
        "top-left",
        "top-right",
        "bottom-left",
        "bottom-right",
      ]),
      floatingProps: PropTypes.object,
      onClick: PropTypes.func,
      onKeyUp: PropTypes.func,
      onKeyDown: PropTypes.func,
      onMouseUp: PropTypes.func,
      onMouseDown: PropTypes.func,
      onMouseLeave: PropTypes.func,
      onTouchStart: PropTypes.func,
      onTouchMove: PropTypes.func,
      onTouchEnd: PropTypes.func,
    };
  } catch (e) {}
}

export default Button;
