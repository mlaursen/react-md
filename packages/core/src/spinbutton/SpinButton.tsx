"use client";

import { type HTMLAttributes, forwardRef } from "react";

import { type LabelRequiredForA11y } from "../types.js";
import { defaultGetSpinButtonTextContent } from "./defaults.js";
import {
  type SpinButtonOptions,
  type SpinButtonUncontrolledValueOptions,
  type SpinButtonValueOptions,
} from "./types.js";
import { useSpinButton } from "./useSpinButton.js";

/**
 * @since 6.4.0
 */
export interface SpinButtonProps
  extends Omit<HTMLAttributes<HTMLDivElement>, keyof SpinButtonOptions>,
    Omit<SpinButtonOptions<HTMLDivElement>, "ref" | "children"> {}

/**
 * The `SpinButton` component can be used as a simple wrapper around the
 * {@link useSpinButton} hook if it should be rendered within a `<div>` and
 * only the rendering behavior is required.
 *
 * @since 6.4.0
 */
export const SpinButton = forwardRef<
  HTMLDivElement,
  LabelRequiredForA11y<SpinButtonProps> & SpinButtonValueOptions
>(function SpinButton(props, ref) {
  const {
    id,
    min,
    max,
    minDigits,
    maxDigits,
    step,
    form,
    error,
    disabled,
    readOnly,
    required,
    fallback,
    mappings,
    value,
    defaultValue,
    onValueChange,
    getValueText,
    getTextContent = defaultGetSpinButtonTextContent,
    placeholderChar,
    defaultKeyboardValue,
    ...remaining
  } = props;

  const valueProps = {
    value,
    defaultValue,
  } as SpinButtonUncontrolledValueOptions<HTMLDivElement>;

  const { value: currentValue, spinButtonProps } = useSpinButton({
    id,
    ref,
    min,
    max,
    minDigits,
    maxDigits,
    step,
    form,
    fallback,
    mappings,
    error,
    disabled,
    readOnly,
    required,
    ...valueProps,
    onValueChange,
    getValueText,
    getTextContent,
    placeholderChar,
    defaultKeyboardValue,
  });

  return (
    <div {...remaining} {...spinButtonProps}>
      {getTextContent({
        min,
        max,
        minDigits,
        maxDigits,
        value: currentValue,
        fallback,
        placeholderChar,
      })}
    </div>
  );
});
