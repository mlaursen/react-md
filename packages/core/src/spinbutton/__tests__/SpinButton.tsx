import { describe, expect, it, jest } from "@jest/globals";
import { type ReactElement, createRef, useState } from "react";

import { Button } from "../../button/Button.js";
import { Form } from "../../form/Form.js";
import {
  type ReactMDRenderOptions,
  render,
  screen,
  userEvent,
} from "../../test-utils/index.js";
import { type MinMaxRange } from "../../types.js";
import { SpinButton, type SpinButtonProps } from "../SpinButton.js";
import { type SpinButtonOptions, type SpinButtonValue } from "../types.js";

const setup = (
  props?: Partial<Omit<SpinButtonProps, "value" | "setValue">>,
  options?: ReactMDRenderOptions
) => {
  const label = props?.["aria-label"] ?? "Label";
  const user = userEvent.setup();
  const { rerender } = render(
    <SpinButton {...props} aria-label={label} />,
    options
  );
  const spinbutton = screen.getByRole("spinbutton", { name: label });

  return {
    user,
    spinbutton,
    rerender: (props?: Partial<Omit<SpinButtonProps, "value" | "setValue">>) =>
      rerender(
        <SpinButton {...props} aria-label={props?.["aria-label"] ?? label} />
      ),
  };
};

describe("SpinButton", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();
    const props = {
      ref,
      "aria-label": "Example",
    } as const;

    const { rerender } = render(<SpinButton {...props} />);

    const spinbutton = screen.getByRole("spinbutton", { name: "Example" });
    expect(spinbutton).toMatchSnapshot();
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(spinbutton);
    expect(spinbutton).toMatchSnapshot();

    rerender(
      <SpinButton
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(spinbutton).toMatchSnapshot();
  });

  it("should allow for a default value", () => {
    const { spinbutton } = setup({ defaultValue: 3 });

    expect(spinbutton).toHaveValue(3);
    expect(spinbutton).toHaveAttribute("aria-valuenow", "3");
    expect(spinbutton).not.toHaveAttribute("aria-valuetext");
    expect(spinbutton).toHaveTextContent("3");
  });

  it("should allow the value to be controlled", async () => {
    const user = userEvent.setup();
    function Test(): ReactElement {
      const [value, setValue] = useState<SpinButtonValue>(null);

      return (
        <>
          <Button onClick={() => setValue(null)}>Set null</Button>
          <Button onClick={() => setValue(3)}>Set 3</Button>
          <SpinButton
            aria-label="Example"
            value={value}
            onValueChange={({ value }) => setValue(value)}
          />
        </>
      );
    }

    render(<Test />);
    const setNull = screen.getByRole("button", { name: "Set null" });
    const set3 = screen.getByRole("button", { name: "Set 3" });
    const spinbutton = screen.getByRole("spinbutton", { name: "Example" });

    expect(spinbutton).not.toHaveAttribute("aria-valuenow");
    expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
    expect(spinbutton).toBeEmptyDOMElement();

    await user.click(set3);
    expect(spinbutton).toHaveValue(3);
    expect(spinbutton).toHaveAttribute("aria-valuenow", "3");
    expect(spinbutton).not.toHaveAttribute("aria-valuetext");
    expect(spinbutton).toHaveTextContent("3");

    await user.click(setNull);
    expect(spinbutton).not.toHaveAttribute("aria-valuenow");
    expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
    expect(spinbutton).toBeEmptyDOMElement();
  });

  it("should update the value as the user types", async () => {
    const { user, spinbutton } = setup();

    expect(spinbutton).not.toHaveAttribute("aria-valuenow");
    expect(spinbutton).toBeEmptyDOMElement();

    await user.type(spinbutton, "3");
    expect(spinbutton).toHaveValue(3);
    expect(spinbutton).toHaveAttribute("aria-valuenow", "3");
    expect(spinbutton).not.toHaveAttribute("aria-valuetext");
    expect(spinbutton).toHaveTextContent("3");

    await user.type(spinbutton, "0", { skipClick: true });
    expect(spinbutton).toHaveValue(30);
    expect(spinbutton).toHaveAttribute("aria-valuenow", "30");
    expect(spinbutton).not.toHaveAttribute("aria-valuetext");
    expect(spinbutton).toHaveTextContent("30");
  });

  it("should ignore leading zeros if the min/minDigits are not provided", async () => {
    const { user, spinbutton } = setup();

    expect(spinbutton).not.toHaveAttribute("aria-valuenow");
    expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
    expect(spinbutton).toBeEmptyDOMElement();

    await user.type(spinbutton, "00000");
    expect(spinbutton).toHaveValue(0);
    expect(spinbutton).toHaveAttribute("aria-valuenow", "0");
    expect(spinbutton).not.toHaveAttribute("aria-valuetext");
    expect(spinbutton).toHaveTextContent("0");

    await user.type(spinbutton, "1", { skipClick: true });
    expect(spinbutton).toHaveValue(1);
    expect(spinbutton).toHaveAttribute("aria-valuenow", "1");
    expect(spinbutton).not.toHaveAttribute("aria-valuetext");
    expect(spinbutton).toHaveTextContent("1");
  });

  it("should allow for a custom getValueText to describe the current value", async () => {
    const { user, spinbutton } = setup({
      getValueText: (value) => `Current value: ${value}`,
    });

    expect(spinbutton).toHaveAttribute("aria-valuetext", "Current value: null");

    await user.type(spinbutton, "12");
    expect(spinbutton).toHaveAttribute("aria-valuetext", "Current value: 12");
  });

  describe("fallback text content", () => {
    it("should be able to display a fallback text value while the value is null", async () => {
      const { user, spinbutton } = setup({ fallback: "HH" });

      expect(spinbutton).not.toHaveValue("aria-valuenow");
      expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
      expect(spinbutton).toHaveTextContent("HH");

      await user.type(spinbutton, "3");
      expect(spinbutton).toHaveValue(3);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "3");
      expect(spinbutton).not.toHaveAttribute("aria-valuetext");
      expect(spinbutton).toHaveTextContent("03");
    });

    it("should default to a fallback of hyphens if a min value was provided", async () => {
      const { spinbutton, rerender, user } = setup({ min: 0 });

      expect(spinbutton).not.toHaveValue("aria-valuenow");
      expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
      expect(spinbutton).toHaveTextContent("-");

      rerender({ min: 10 });

      expect(spinbutton).not.toHaveValue("aria-valuenow");
      expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
      expect(spinbutton).toHaveTextContent("--");

      await user.type(spinbutton, "2");
      expect(spinbutton).not.toHaveValue("aria-valuenow");
      expect(spinbutton).toHaveAttribute("aria-valuetext", "No value selected");
      expect(spinbutton).toHaveTextContent("02");

      await user.type(spinbutton, "0", { skipClick: true });
      expect(spinbutton).toHaveValue(20);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "20");
      expect(spinbutton).not.toHaveAttribute("aria-valuetext");
      expect(spinbutton).toHaveTextContent("20");
    });

    it("should allow for a custom getTextContent customize the fallback text content", async () => {
      const { user, spinbutton } = setup({
        getTextContent: (options) => {
          const { value } = options;
          if (value === null) {
            return "Nothing";
          }

          return `${value}`;
        },
      });

      expect(spinbutton).not.toHaveAttribute("aria-valuenow");
      expect(spinbutton).toHaveTextContent("Nothing");
      await user.type(spinbutton, "12");

      expect(spinbutton).toHaveAttribute("aria-valuenow", "12");
      expect(spinbutton).toHaveTextContent("12");

      await user.keyboard("{Backspace}");
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");
      expect(spinbutton).toHaveTextContent("Nothing");
    });
  });

  describe("keyboard overrides", () => {
    it("should prevent default beahvior for ArrowLeft, ArrowRight, Home, and End keys to prevent the cursor from moving in the contentEditable element", async () => {
      const { user, spinbutton } = setup({ defaultValue: 10 });

      await user.type(spinbutton, "{ArrowLeft}");
      expect(spinbutton).toHaveSelection("10");

      await user.type(spinbutton, "{ArrowRight}");
      expect(spinbutton).toHaveSelection("10");

      await user.type(spinbutton, "{Home}");
      expect(spinbutton).toHaveSelection("10");

      await user.type(spinbutton, "{End}");
      expect(spinbutton).toHaveSelection("10");
    });

    it("should allow the Home key to jump to the min value", async () => {
      const { user, spinbutton } = setup({ min: 0, defaultValue: 10 });

      await user.type(spinbutton, "{Home}");
      expect(spinbutton).toHaveValue(0);
      expect(spinbutton).toHaveSelection("0");
    });

    it("should allow the End key to jump to the max value", async () => {
      const { user, spinbutton } = setup({ max: 20, defaultValue: 10 });

      await user.type(spinbutton, "{End}");
      expect(spinbutton).toHaveValue(20);
      expect(spinbutton).toHaveSelection("20");
    });

    it("should attempt to submit the form when the enter key is pressed", async () => {
      const onSubmit = jest.fn();
      const { user, spinbutton } = setup(
        {},
        {
          wrapper: ({ children }) => (
            <Form onSubmit={onSubmit}>
              {children}
              <Button type="submit">Submit</Button>
            </Form>
          ),
        }
      );

      await user.type(spinbutton, "3{Enter}");
      expect(onSubmit).toHaveBeenCalledTimes(1);
    });
  });

  describe("keyboard increment", () => {
    it("should allow the user to increment the value with the ArrowUp key", async () => {
      const { user, spinbutton } = setup({ min: 0, max: 3, defaultValue: 0 });
      expect(spinbutton).toHaveValue(0);

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(1);

      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(2);

      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(3);
    });

    it("should loop the value to the min value if it was provided", async () => {
      const { user, spinbutton } = setup({ min: 0, max: 3, defaultValue: 3 });
      expect(spinbutton).toHaveValue(3);

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(0);
    });

    it("should stop at the max value if a min was not provided", async () => {
      const { user, spinbutton } = setup({ max: 3, defaultValue: 3 });
      expect(spinbutton).toHaveValue(3);

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(3);
    });

    it("should allow the initial increment value to start from the defaultKeyboardValue if there is no value in the spinbutton", async () => {
      const { user, spinbutton } = setup({ defaultKeyboardValue: 3 });
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(4);
    });

    it("should set the initial increment value to the min if it was provided without a defaultKeyboardValue", async () => {
      const { user, spinbutton } = setup({ min: 2 });
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(3);
    });

    it("should set the initial increment value to the max if it was provided without a defaultKeyboardValue and min", async () => {
      const { user, spinbutton } = setup({ max: 9 });
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(9);
    });

    it("should set the initial increment value to 0 if defaultKeyboardValue, min, and max were not provided", async () => {
      const { user, spinbutton } = setup();
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(1);
    });
  });

  describe("keyboard decrement", () => {
    it("should allow the user to decrement the value with the ArrowDown key", async () => {
      const { user, spinbutton } = setup({ min: 0, max: 3, defaultValue: 3 });

      expect(spinbutton).toHaveValue(3);

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(2);

      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(1);

      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(0);
    });

    it("should loop the value to the max value if it was provided", async () => {
      const { user, spinbutton } = setup({ min: 0, max: 3, defaultValue: 0 });
      expect(spinbutton).toHaveValue(0);

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(3);
    });

    it("should stop at the max value if a max was not provided", async () => {
      const { user, spinbutton } = setup({ min: 3, defaultValue: 3 });
      expect(spinbutton).toHaveValue(3);

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(3);
    });

    it("should allow the initial decrement value to start from the defaultKeyboardValue if there is no value in the spinbutton", async () => {
      const { user, spinbutton } = setup({ defaultKeyboardValue: 3 });
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(2);
    });

    it("should set the initial decrement value to the min if it was provided without a defaultKeyboardValue", async () => {
      const { user, spinbutton } = setup({ min: 2 });
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(2);
    });

    it("should set the initial decrement value to the max if it was provided without a defaultKeyboardValue and min", async () => {
      const { user, spinbutton } = setup({ max: 9 });
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(8);
    });

    it("should set the initial decrement value to 0 if defaultKeyboardValue, min, and max were not provided", async () => {
      const { user, spinbutton } = setup();
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      spinbutton.focus();
      await user.keyboard("{ArrowDown}");
      expect(spinbutton).toHaveValue(-1);
    });
  });

  describe("form states", () => {
    it("should add the correct attributes when disabled and prevent any changes", async () => {
      const { user, spinbutton } = setup({
        min: 0,
        max: 10,
        disabled: true,
        defaultValue: 5,
      });

      expect(spinbutton).toHaveAttribute("aria-disabled", "true");
      expect(spinbutton).not.toHaveAttribute("tabIndex");
      expect(spinbutton).toHaveValue(5);

      await user.click(spinbutton);
      expect(spinbutton).not.toHaveFocus();
    });

    it("should add the correct attributes when readOnly and prevent any changes", async () => {
      const { user, spinbutton } = setup({
        min: 0,
        max: 10,
        readOnly: true,
        defaultValue: 5,
      });

      expect(spinbutton).toHaveAttribute("aria-readonly", "true");
      expect(spinbutton).toHaveAttribute("tabIndex", "0");
      expect(spinbutton).toHaveValue(5);

      await user.click(spinbutton);
      expect(spinbutton).toHaveFocus();

      await user.keyboard("{ArrowUp}");
      expect(spinbutton).toHaveValue(5);

      await user.keyboard("3");
      expect(spinbutton).toHaveValue(5);
    });

    it("should flag the aria-required state when errored", () => {
      const { spinbutton } = setup({
        min: 0,
        max: 10,
        defaultValue: 12,
        required: true,
      });

      expect(spinbutton).toHaveAttribute("aria-required", "true");
    });

    it("should flag the aria-invalid state when errored", () => {
      const { spinbutton } = setup({
        min: 0,
        max: 10,
        defaultValue: 12,
        // this would normally be set to true only when the form is submitted
        error: true,
      });

      expect(spinbutton).toHaveAttribute("aria-invalid", "true");
    });
  });

  describe("typing", () => {
    const HOUR_RANGE: MinMaxRange = {
      min: 1,
      max: 12,
    };

    const HOUR_SETUP = {
      ...HOUR_RANGE,
      fallback: "HH",
    } satisfies Partial<SpinButtonOptions>;

    it("should ignore anything that is not a number", async () => {
      const { user, spinbutton } = setup(HOUR_SETUP);

      expect(spinbutton).toHaveTextContent("HH");
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      await user.type(spinbutton, "hello, world!");
      expect(spinbutton).not.toHaveAttribute("aria-valuenow");
      expect(spinbutton).toHaveSelection("HH");

      await user.keyboard("1");
      expect(spinbutton).toHaveValue(1);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "1");
      expect(spinbutton).toHaveSelection("01");

      // ignored keys are not counted with the keypress count
      await user.keyboard("-3");
      expect(spinbutton).toHaveValue(1);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "1");
      expect(spinbutton).toHaveSelection("01");

      await user.keyboard("-3");
      expect(spinbutton).toHaveValue(3);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "3");
      expect(spinbutton).toHaveSelection("03");
    });

    it("should allow for custom key mappings to support the AM/PM sort of spin buttons", async () => {
      const { user, spinbutton } = setup({
        min: 0,
        max: 1,
        mappings: { a: 0, p: 1 },
      });

      expect(spinbutton).not.toHaveAttribute("aria-valuenow");

      await user.type(spinbutton, "a");
      expect(spinbutton).toHaveValue(0);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "0");

      await user.keyboard("p");
      expect(spinbutton).toHaveValue(1);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "1");

      await user.keyboard("aaaaaafjdks");
      expect(spinbutton).toHaveValue(0);
      expect(spinbutton).toHaveAttribute("aria-valuenow", "0");
    });
  });

  describe("onValueChange", () => {
    it("should call onValueChange when typing", async () => {
      const onValueChange = jest.fn();
      const { user, spinbutton } = setup({
        onValueChange,
        min: 0,
        max: 10,
        defaultValue: 5,
      });

      await user.type(spinbutton, "1");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "type",
        value: 1,
      });

      await user.keyboard("1");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "typed-to-completion",
        // see resolveInputEvent tests for why this is 1
        value: 1,
      });
      expect(onValueChange).toHaveBeenCalledTimes(2);
    });

    it("should call onValueChange when clearing the value", async () => {
      const onValueChange = jest.fn();
      const { user, spinbutton } = setup({
        onValueChange,
        min: 0,
        max: 10,
        defaultValue: 5,
      });

      await user.type(spinbutton, "{Backspace}");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "cleared",
        value: null,
      });
    });

    it("should call onValueChange for keyboard jumps", async () => {
      const onValueChange = jest.fn();
      const { user, spinbutton } = setup({
        onValueChange,
        min: 0,
        max: 10,
        defaultValue: 5,
      });

      expect(spinbutton).toHaveValue(5);
      expect(onValueChange).not.toHaveBeenCalled();

      await user.type(spinbutton, "{ArrowUp}");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "change",
        value: 6,
      });

      await user.keyboard("{ArrowDown}");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "change",
        value: 5,
      });

      await user.keyboard("{Home}");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "change",
        value: 0,
      });

      await user.keyboard("{End}");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "change",
        value: 10,
      });
    });

    it("should not call the onValueChange for ignored events", async () => {
      const onValueChange = jest.fn();
      const { user, spinbutton } = setup({
        onValueChange,
        min: 0,
        max: 10,
        defaultValue: 5,
      });

      await user.type(spinbutton, "hello, world! let's type some stuff");
      expect(onValueChange).not.toHaveBeenCalled();
    });

    it("should not call the onValueChange for placeholder digits", async () => {
      const onValueChange = jest.fn();
      const { user, spinbutton } = setup({
        onValueChange,
        min: 1900,
        max: 2100,
      });

      await user.type(spinbutton, "190");
      expect(onValueChange).not.toHaveBeenCalled();

      await user.keyboard("1");
      expect(onValueChange).toHaveBeenCalledWith({
        event: expect.any(Object),
        reason: "typed-to-completion",
        value: 1901,
      });
    });
  });
});
