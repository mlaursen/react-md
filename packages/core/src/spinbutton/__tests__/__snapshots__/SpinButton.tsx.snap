// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinButton should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<div
  aria-label="Example"
  aria-valuetext="No value selected"
  autocapitalize="none"
  autocorrect="off"
  contenteditable="true"
  id="spinbutton-:r0:"
  inputmode="numeric"
  role="spinbutton"
  spellcheck="false"
  tabindex="0"
/>
`;

exports[`SpinButton should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<div
  aria-label="Example"
  aria-valuetext="No value selected"
  autocapitalize="none"
  autocorrect="off"
  contenteditable="true"
  id="spinbutton-:r0:"
  inputmode="numeric"
  role="spinbutton"
  spellcheck="false"
  tabindex="0"
/>
`;

exports[`SpinButton should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<div
  aria-label="Example"
  aria-valuetext="No value selected"
  autocapitalize="none"
  autocorrect="off"
  class="custom-class-name"
  contenteditable="true"
  id="spinbutton-:r0:"
  inputmode="numeric"
  role="spinbutton"
  spellcheck="false"
  style="color: white;"
  tabindex="0"
/>
`;
